### PR TITLE
fix(@aws-amplify/analytics): unlink user id from endpoint when limit exceeds

### LINKED ISSUE #1378: Pinpoint Analytics Exceeded maximum endpoint per user count 10
Do you want to request a feature or report a bug?
seems like a misconfiguration issue? 

What is the current behavior?

Amplify analytics configuration
```
Analytics: {
    appId: '***',
    region: 'us-east-1',
    // endpointId: '***'
  }
```
When I had endpointId as a constant, I could only see one endpoint at Pinpoint. Thus, I got advice in gitter that I should let Amplify set the endpointId. Thus, I did that and I see this `Exceeded maximum endpoint per user count 10` error for some users:
```
400 Bad Request with pinpoint saying that
{"RequestID":"*","Message":"Exceeded maximum endpoint per user count 10"
```

What is the expected behavior?

Which versions of Amplify, and which browser / OS are affected by this issue? Did this work in previous versions?
0.4.3

You can turn on the debug mode to provide more info for us by setting ```window.LOG_LEVEL = 'DEBUG';``` in your app.
yes

### PR DESCRIPTION
**Pull request by powerful23**
*Issue #, if available:*
fixes #1378 
*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/analytics/src/Providers/AWSPinpointProvider.ts [hunk 1]
```diff
@@ -379,15 +379,79 @@ export default class AWSPinpointProvider implements AnalyticsProvider {
             that.pinpointClient.updateEndpoint(update_params, (err, data) => {
                 if (err) {
                     logger.debug('updateEndpoint failed', err);
-                    res(false);
+                    if (err.message === 'Exceeded maximum endpoint per user count 10') {
```
Comment (haverchuck) [inline packages/analytics/src/Providers/AWSPinpointProvider.ts]: Will '10' always be the threshhold in this message?

Comment (powerful23) [reply packages/analytics/src/Providers/AWSPinpointProvider.ts]: @haverchuck it should be fixed. Also there is no error code or something that we can tell which error type is.

**Comment by janpapenbrock** [comment]
This requires some policy additions, I assume? Using the latest release `1.1.18`, I see a lot of 403ing requests to `https://pinpoint.us-east-1.amazonaws.com/v1/apps/{some-id}/users/eu-central-1%3A{another-id}`. As this is running with user credentials in the frontend, can a user only list and delete their own endpoints?

**Comment by powerful23** [comment]
@janpapenbrock it can be used to get other users' endpoints as long as you pass the corresponding user-id.

**Comment by mauliksoni110** [comment]
Please correct me if I am wrong here, but the line
this._removeUnusedEndpoints(appId, credentials.identityId)
is trying to remove the endpoint of user using `credentials.identityId`,
but we do the manual `.updateEndpoint` using our Mongo-id (for example `5901dd7c72454d001a026bc3`), and thus, `this._removeUnusedEndpoints` never removes endpoints for userId `5901dd7c72454d001a026bc3`, but tries to remove it for credentials.identityId (`us-east-1-6c********def`).

So the user `5901dd7c72454d001a026bc3` keeps getting the 'Maximum endpoint per user....' error.

Any way to avoid this but keep using mongo-id `5901dd7c72454d001a026bc3` for UserId key in updateEndpoint?

**Comment by powerful23** [comment]
@mauliksoni110 thanks for point out, will send a pr to fix this.

**Pull request merged**